pkg/reconciler/delivery/delivery.go

```diff
@@ -99,17 +109,53 @@ func (c *Reconciler) updateRoute(ctx context.Context, cfg *v1.Configuration) err
 	if err != nil {
 		return err
 	}
+
 	logger.Info("Applying updated Route object")
 	_, err = c.client.ServingV1().Routes(cfg.Namespace).Update(route)
 
+	// when the percentages reach 0/100, we know that we don't need to queue future events
+	if *route.Spec.Traffic[0].Percent != 0 {
+		logger.Infof("Queueing event for %v seconds later", policy.Threshold)
+		c.followup(cfg, time.Duration(policy.Threshold) * time.Second)
+	} else {
+		logger.Info("Progressive rollout completed!")
+	}
+
 	return err
 }
 
 // isRouteStatusUpToDate determines if the current Route status already matches our desired state
 func isRouteStatusUpToDate(route *v1.Route, newRevName string) bool {
-	for idx := range route.Status.Traffic {
-		if route.Status.Traffic[idx].RevisionName == newRevName {
-			return true
+	// the Route status is up to date if:
+	// 1. the new Revision is listed in the status traffic targets, AND
+	// 2. the status IngressReady condition's last transition time stamp hasn't expired
+	// OR if:
+	// 3. the new Revision is listed in the status traffic targets, AND
+	// 4. the new Revision already reached 100%
+	var nameListed bool = false
+	for _, t := range route.Status.Traffic {
+		if t.RevisionName == newRevName {
+			nameListed = true
+			break
+		}
+	}
+	if !nameListed {
+		return false
+	}
+	if len(route.Status.Traffic) == 1 || *route.Status.Traffic[1].Percent == 100 {
+		return true
+	}
+	for _, c := range route.Status.Conditions {
+		if c.Type != "IngressReady" {
+			continue
+		}
+		// this is the tricky part, using time stamps to figure out if Route is up to date
+		// IngressReady transition has a short delay of a couple seconds, which demands some wiggle room
+		// this sacrifices some timing precision, but I haven't thought of a better way yet
+		// it also poses problems when the user applies the 2nd Revision very soon (fewer than policy.Threshold
```

Comment:
Is this related to `Unsupported use case: current implementation only supports 2 Revisions at once`?

Reply:
No, the problem here is that after I do `c.client.ServingV1().Routes(cfg.Namespace).Update(route)` it takes a few seconds for the last transition timestamp to be updated. For example, if the above code is executed at t=0s, then the last transition timestamp would probably be t=5s. However, `AddAfter` is called almost immediately after `c.client.ServingV1().Routes(cfg.Namespace).Update(route)`, so when the next time ReconcileKind is invoked due to the timed event, not enough time would have elapsed compared to the last transition timestamp. In this example, suppose the time interval is 20 seconds, and we do `AddAfter(20 * time.Second)` then ReconcileKind will be invoked at t=21s, which is only 16s after the last transition timestamp. Thus, `isRouteStatusUpToDate` cannot naively compare the the last transition timestamp against the time interval. In this case if it compares 16 with 20 it will conclude that the timestamp hasn't expired, and will fail to trigger a Route update even if it should.

Another problem is that when we initially apply the first Revision (R1), it will still trigger `AddAfter`. For example, if the user applies R1 at t=1s, then the reconciler will enqueue an event at t=21s. However, if the user applies R2 at any time between t=1 and t=21, the event at t=21s might prematurely cause the progressive rollout to move on to the next stage. Now that I think about it, this problem might be avoided by preventing R1 from triggering `AddAfter`, and I will fix it and commit a new version.

That being said, I don't think accommodating > 2 Revisions would exacerbate or alleviate either of these problems.

Reply:
The second problem is now resolved. I prevented R1 from triggering `AddAfter`. However, the first problem is still present.